---
title: "Understanding the transcriptional response to ER stress in Chinese hamster ovary cells using multiplexed single cell RNA-seq"
output: html_notebook
---

## Introduction


## Prepare for analysis
### Load libraries & required functions
```{r include=FALSE}
package_list <- c(
  "DropletUtils", "Matrix", "tidyverse", "readxl", "scales", "ggridges", "WGCNA", "scWGCNA",
  "Seurat", "viridis", "fuzzyjoin", "WebGestaltR", "ggpubr", "patchwork", "scater"
)

lapply(package_list, require, character.only = TRUE)

set.seed(38)
```

## Create results directory
```{r}
results_dir <- "../results/"
if (!dir.exists(results_dir)) {
  dir.create(results_dir, recursive = TRUE)
}
```

# qPCR of TM treatment
```{r fig.height=5, fig.width=7}
qpcr_data <- read_excel("../data/qpcr_analysis.xlsx")

qpcr_data <- qpcr_data %>%
  pivot_longer(cols = c("Hspa5", "Ddit3", "Atf4", "sXbp1")) %>%
  mutate(fold_change = signif(value, 2)) %>%
  mutate(Condition_edit = case_when(
    Condition == "TM-" ~ "control",
    Condition == "TM+ 1hr" ~ "1hr",
    Condition == "TM+ 2hr" ~ "2hr",
    Condition == "TM+ 4hr" ~ "4hr",
    Condition == "TM+ 8hr" ~ "8hr"
  ))

col_pal_vidiris <- c("#fde725", "#5ec962", "#21918c", "#31688e", "#440154")

tm_qpcr_plot <- qpcr_data %>%
  ggboxplot(
    x = "Condition_edit", y = "fold_change",
    fill = "Condition_edit", palette = col_pal_vidiris
  ) +
  geom_point(
    data = qpcr_data, aes(x = Condition_edit, y = fold_change),
    position = position_jitter(width = 0), size = 0.2, color = "grey"
  ) +
  labs(x = "", y = "Fold change") +
  facet_wrap(~name, scales = "free_y", nrow = 1) +
  theme_bw() +
  theme(
    legend.position = "none",
    strip.text.x = element_text(face = "bold.italic"),
    strip.background = element_blank(),
    panel.spacing = unit(1, "lines"),
    axis.text.x = element_text(angle = 35, vjust = 1, hjust = 1)
  ) +
  # axis.text.x = element_text(angle=35, hjust=1, size = 8)) +
  scale_y_continuous(breaks = pretty_breaks(n = 4, min.n = 0))

tm_qpcr_plot
ggsave(plot = tm_qpcr_plot, filename = paste(results_dir, "Figure 1A.png", sep = ""), width = 7, height = 2, device = "png", dpi = 700)
```
# Single cell RNA-seq analysis
## Preparation
### Load functions
Importing counts from Kalisto | Bustools as described in the [tutorial] (https://www.kallistobus.tools/tutorials/kb_getting_started/r/kb_intro_2_r/)
```{r include=FALSE}
# import counts
read_count_output <- function(dir, name) {
  dir <- normalizePath(dir, mustWork = TRUE)
  m <- readMM(paste0(dir, "/", name, ".mtx"))
  m <- Matrix::t(m)
  m <- as(m, "dgCMatrix")
  # The matrix read has cho.combined.cds in rows
  ge <- ".genes.txt"
  genes <- readLines(file(paste0(dir, "/", name, ge)))
  barcodes <- readLines(file(paste0(dir, "/", name, ".barcodes.txt")))
  colnames(m) <- barcodes
  rownames(m) <- genes
  return(m)
}

# Knee plot for filtering empty droplets
knee_plot <- function(bc_rank) {
  knee_plt <- tibble(
    rank = bc_rank[["rank"]],
    total = bc_rank[["total"]]
  ) %>%
    distinct() %>%
    dplyr::filter(total > 0)
  annot <- tibble(
    inflection = metadata(bc_rank)[["inflection"]],
    rank_cutoff = max(bc_rank$rank[bc_rank$total > metadata(bc_rank)[["inflection"]]])
  )
  p <- ggplot(knee_plt, aes(total, rank)) +
    geom_line() +
    geom_hline(aes(yintercept = rank_cutoff), data = annot, linetype = 2) +
    geom_vline(aes(xintercept = inflection), data = annot, linetype = 2) +
    scale_x_log10() +
    scale_y_log10() +
    annotation_logticks() +
    labs(y = "Rank", x = "Total UMIs")
  return(p)
}

FilterGenes <- function(object, min.value = 1, min.cells = 0, genes = NULL) {
  # parameters.to.store <- as.list(environment(), all = TRUE)[names(formals("FilterGenes"))]
  # object <- Seurat:::SetCalcParams(object = object, calculation = "FilterGenes", ... = parameters.to.store)
  genes.use <- rownames(object@assays$RNA@counts)

  if (!is.null(genes)) {
    genes.use <- intersect(genes.use, genes)
    object@assays$RNA@counts <- object@assays$RNA@counts[genes.use, ]
    return(object)
  } else if (min.cells > 0) {
    num.cells <- Matrix::rowSums(object@assays$RNA > min.value)
    genes.use <- names(num.cells[which(num.cells >= min.cells)])
    object@assays$RNA <- object@assays$RNA[genes.use, ]
    return(object)
  } else {
    return(object)
  }
}
```

## mRNA libraries

### Set fixed filtering thresholds
here we set values for different criteria using in the analysis
```{r}
min_cells_with_expression <- 50
min_genes_per_cell <- 200
mt_threshold <- 10
```

### load genelists
```{r include=FALSE}
# cell cycle genes
cgr_s_phase <- read.csv(file = "../data/s_phase_genes.csv")$Seurat.gene.ID

cgr_g2m_phase <- read.csv(file = "../data/g2m_phase_genes.csv")$Seurat.gene.ID

# mitochondrial genes
mito <- c(
  "ND1", "ND2", "COX1", "COX2", "ATP8", "ATP6",
  "COX3", "ND3", "ND4L", "ND4", "ND5", "ND6", "CYTB"
)
```


## Library 1

### Load
```{r warning=FALSE}
# import KB counts
# tagged_cells_matrix_1 <- read_count_output("../../kallisto_counts/multiplexed_scRNAseq_data_1/counts_unfiltered/",
#                                            name = "cells_x_genes")
#
# tot_counts <- Matrix::colSums(tagged_cells_matrix_1)
# summary(tot_counts)
#
# bc_rank <- barcodeRanks(tagged_cells_matrix_1, lower = 100)
#
# options(repr.plot.width=9, repr.plot.height=6)
# knee_plot(bc_rank)
# tagged_cells_matrix_1 <- tagged_cells_matrix_1[, tot_counts > metadata(bc_rank)$inflection]
# #tagged_cells_matrix_1 <- tagged_cells_matrix_1[Matrix::rowSums(tagged_cells_matrix_1) > 0,]
# dim(tagged_cells_matrix_1)
#
# # create a seurat object
# cell_library_1 <- CreateSeuratObject(counts = tagged_cells_matrix_1,
#                                         min.cells = min_cells_with_expression,
#                                         min.features = min_genes_per_cell)

# line to be removed when code finished
cell_library_1 <- readRDS("~/GitHub/multiplexed_CHO_scrnaseq/data/tagged_cells_lib1.rds")

cell_library_1
```

### Preprocess
```{r warning=FALSE}
# determine the number of umis originating from each cell
cell_library_1[["percent.mt"]] <- PercentageFeatureSet(cell_library_1, features = mito)

# remove the mito genes from further analyses
keep <- c(!rownames(cell_library_1) %in% mito)
cell_library_1 <- subset(x = cell_library_1, features = c(1:(dim(cell_library_1)[1]))[keep])

# filter cells based on number of umis, genes and % mtdna
qc.nCount_RNA <- isOutlier(cell_library_1$nCount_RNA, nmads = 3, log=TRUE, type="both")
qc.nFeature_RNA  <- isOutlier(cell_library_1$nFeature_RNA, nmads = 3, log=TRUE, type="both")

count_low <- attr(qc.nCount_RNA , "thresholds")[1]
count_high <- attr(qc.nCount_RNA , "thresholds")[2]

gene_low <- attr(qc.nFeature_RNA , "thresholds")[1]
gene_high <- attr(qc.nFeature_RNA , "thresholds")[2]

cell_lib_1_threshold_line_values <- tibble(
  name = c("# UMIs", "# Genes", "% mito"),
  library = c("Cell library 1", "Cell library 1", "Cell library 1"),
  lower = c(count_low, gene_low, NA),
  upper = c(count_high, gene_high, mt_threshold)
)

cell_library_1_meta_data <- cell_library_1@meta.data %>%
  dplyr::rename(
    `# UMIs` = nCount_RNA,
    `# Genes` = nFeature_RNA,
    `% mito` = percent.mt
  ) %>%
  mutate(library = "Cell library 1")

cell_library_1 <- subset(cell_library_1,
  subset = nFeature_RNA >= gene_low  &
    nFeature_RNA <= gene_high  &
    nCount_RNA > count_low &
    nCount_RNA < count_high &
    percent.mt < mt_threshold
)

cell_library_1@meta.data$library <- rep("library 1", dim(cell_library_1)[2])
cell_library_1
```

## Library 2

### Load
```{r warning=FALSE}
# tagged_cells_matrix_2 <- read_count_output("../../kallisto_counts/multiplexed_scRNAseq_data_2/counts_unfiltered/",
#                                            name = "cells_x_genes")

# tot_counts <- Matrix::colSums(tagged_cells_matrix_2)
# summary(tot_counts)
#
# bc_rank <- barcodeRanks(tagged_cells_matrix_2, lower = 100)
#
# options(repr.plot.width=9, repr.plot.height=6)
# knee_plot(bc_rank)
# tagged_cells_matrix_2 <- tagged_cells_matrix_2[, tot_counts > metadata(bc_rank)$inflection]
# #tagged_cells_matrix_1 <- tagged_cells_matrix_1[Matrix::rowSums(tagged_cells_matrix_1) > 0,]
# dim(tagged_cells_matrix_2)
#
# # create a seurat object
# tagged_cells_lib2 <- CreateSeuratObject(counts = tagged_cells_matrix_2,
#                                         min.cells = min_cells_with_expression,
#                                         min.features = min_genes_per_cell)

cell_library_2 <- readRDS("~/GitHub/multiplexed_CHO_scrnaseq/data/tagged_cells_lib2.rds")

cell_library_2
```

### Preprocess
```{r warning=FALSE}
# determine the number of umis originating from each cell
cell_library_2[["percent.mt"]] <- PercentageFeatureSet(cell_library_2, features = mito)

# remove the mito genes from further analyses
keep <- c(!rownames(cell_library_2) %in% mito)
cell_library_2 <- subset(x = cell_library_2, features = c(1:(dim(cell_library_2)[1]))[keep])

# eliminate cells based on number of umis, genes and % mtdna
qc.nCount_RNA <- isOutlier(cell_library_2$nCount_RNA, nmads=3, log=TRUE, type="both")
qc.nFeature_RNA  <- isOutlier(cell_library_2$nFeature_RNA, nmads=3, log=TRUE, type="both")

count_low <- attr(qc.nCount_RNA , "thresholds")[1]
count_high <- attr(qc.nCount_RNA , "thresholds")[2]

gene_low <- attr(qc.nFeature_RNA , "thresholds")[1]
gene_high <- attr(qc.nFeature_RNA , "thresholds")[2]


cell_lib_2_threshold_line_values <- tibble(
  name = c("# UMIs", "# Genes", "% mito"),
  library = c("Cell library 2", "Cell library 2", "Cell library 2"),
  lower = c(count_low, gene_low, NA),
  upper = c(count_high, gene_high, mt_threshold)
)

cell_library_2_meta_data <- cell_library_2@meta.data %>%
  dplyr::rename(
    `# UMIs` = nCount_RNA,
    `# Genes` = nFeature_RNA,
    `% mito` = percent.mt
  ) %>%
  mutate(library = "Cell library 2")

cell_library_2 <- subset(cell_library_2,
  subset = nFeature_RNA >= gene_low &
    nFeature_RNA <= gene_high  &
    nCount_RNA > count_low &
    nCount_RNA < count_high &
    percent.mt < mt_threshold
)

cell_library_2@meta.data$library <- rep("library 2", dim(cell_library_2)[2])
cell_library_2
```

```{r fig.height=5, fig.width=6, message=FALSE, warning=FALSE}
threshold_line_values <- bind_rows(cell_lib_1_threshold_line_values, cell_lib_2_threshold_line_values)

figure_s2 <- bind_rows(cell_library_1_meta_data, cell_library_2_meta_data) %>%
  pivot_longer(cols = c(`# UMIs`, `# Genes`, `% mito`)) %>%
  ggplot(aes(x = orig.ident, y = value, fill = name)) +
  geom_violin() +
  scale_fill_manual(values = c("#56B4E9", "#E69F00", "#009E73")) +
  geom_jitter(size = 0.1, alpha = 0.1, color = "black") +
  facet_wrap(library ~ name, scales = "free") +
  theme_bw() +
  labs(x = "", y = "") +
  theme(
    panel.spacing = unit(1.5, "lines"),
    axis.text.x = element_blank(),
    strip.background = element_blank(),
    legend.position = "none"
  ) +
  geom_hline(data = threshold_line_values, aes(yintercept = c(lower)), linetype = 11, size = 0.3) +
  geom_hline(data = threshold_line_values, aes(yintercept = c(upper)), linetype = 11, size = 0.3)

figure_s2
ggsave(plot = figure_s2, filename = paste(results_dir, "Figure S1.png", sep = ""), width = 7, height = 5, device = "png", dpi = 700)
```

# Integrate the cells and SBO tags
### Import SBO libaries 
```{r}
# import the SBO data from CITE-seq count
# sbo_lib_1.data <- Read10X(data.dir = "../../CITE-seq_counts/SBO_library_1/SBO_library_1_counts/umi_count/",gene.column = 1)
# sbo_lib_1.data <- sbo_lib_1.data[-6,]

# sbo_lib_2.data <- Read10X(data.dir = "../../CITE-seq_counts/SBO_library_2/SBO_library_2_counts/umi_count/",gene.column = 1)
# sbo_lib_2.data <- sbo_lib_2.data[-6,]

```


## Cell library 1
### Integrate
```{r cell library 1 intergration, message=FALSE, warning=FALSE, include=FALSE}
sbo_lib_1.data <- readRDS("~/GitHub/multiplexed_CHO_scrnaseq/data/sbo_lib_1.data.rds")

tagged_cells_seurat_matrix1 <- as.matrix(GetAssayData(object = cell_library_1, 
                                                      slot = "counts"))

# overlap the cell labels from the cellular and SBO libraries
cho_lib1_joint_bcs <- intersect(colnames(sbo_lib_1.data), 
                                colnames(tagged_cells_seurat_matrix1))


paste(length(cho_lib1_joint_bcs) ," matched for mRNA and SBO library 1" )

# retain the cells that overlap
cho_cells_lib_1 <- tagged_cells_seurat_matrix1[, cho_lib1_joint_bcs]


# create new seurat object
cho_cell_sbo_lib_1 <- CreateSeuratObject(
  counts = cho_cells_lib_1, 
  project = "RNA library 1"
)

cho_cell_sbo_lib_1[["percent.mt"]] <- cell_library_1$percent.mt[cho_lib1_joint_bcs]

cho_cell_sbo_lib_1@meta.data$library <- rep("library 1", dim(cho_cell_sbo_lib_1)[2])

# add a new assay to hold the SBO data
cho_lib_1_sbo <- as.matrix(sbo_lib_1.data[, cho_lib1_joint_bcs]) # intersect

cho_cell_sbo_lib_1[["SBO"]] <- CreateAssayObject(counts = cho_lib_1_sbo)
```

### Classify
```{r}
# normalise
cho_cell_sbo_lib_1 <- NormalizeData(cho_cell_sbo_lib_1, assay = "SBO", 
                               normalization.method = "CLR")

eval_pq_lib1 = tibble()
for (i in seq(from=0.9, to=0.9999999, by =0.005)){

  cho_cell_sbo_lib_1 <- HTODemux(cho_cell_sbo_lib_1, assay = "SBO", 
                          positive.quantile = i)

  hto_demux_classification <- cho_cell_sbo_lib_1@meta.data %>%
    filter(SBO_classification.global != "Negative") %>%
    group_by(SBO_classification.global) %>%
    summarise(cells = n()) %>%
    mutate(threshold=i)
  
  eval_pq_lib1 <- bind_rows(eval_pq_lib1, hto_demux_classification) %>%
    mutate(library="library 1")
  }

eval_pq_lib1_plot <- eval_pq_lib1 %>%
  ggplot(aes(x=threshold, y=cells, color=SBO_classification.global)) +
           geom_point() +
           geom_line() + 
  geom_vline(xintercept = 0.975, linetype=11, size=0.2) +
  theme_bw() +
  labs(y="# Cells", x="HTODemux positive quantile treshold", color="SBO identification", title="SBO library 1")
cho_cell_sbo_lib_1 <- HTODemux(cho_cell_sbo_lib_1, assay = "SBO", 
                          positive.quantile = 0.975)


# show table
table(cho_cell_sbo_lib_1$SBO_maxID, cho_cell_sbo_lib_1$SBO_classification.global)

# data for merged classification plot
classification_library_1 <- data.frame(table(cho_cell_sbo_lib_1$SBO_maxID, cho_cell_sbo_lib_1$SBO_classification.global)) %>%
  mutate(library="RNA library 1")

# remove cells with no matched SBO 
cho_cell_sbo_lib_1.subset <- subset(cho_cell_sbo_lib_1, idents = "Negative", invert = T)
```

## Cell library 2
### Integrate
```{r message=FALSE, warning=FALSE, include=FALSE}
sbo_lib_2.data <- readRDS("~/GitHub/multiplexed_CHO_scrnaseq/data/sbo_lib_2.data.rds")

tagged_cells_seurat_matrix2 <- as.matrix(GetAssayData(object = cell_library_2, 
                                                      slot = "counts"))

# overlap the cell labels from the cellular and SBO libraries
cho_lib2_joint_bcs <- intersect(colnames(sbo_lib_2.data), 
                                colnames(tagged_cells_seurat_matrix2))

paste(length(cho_lib2_joint_bcs) ," matched for mRNA and SBO library 2" )

# retain the cells that overlap
cho_cells_lib_2 <- tagged_cells_seurat_matrix2[, cho_lib2_joint_bcs]


# create new seurat object
cho_cell_sbo_lib_2 <- CreateSeuratObject(
  counts = cho_cells_lib_2, 
  project = "RNA library 2"
)

cho_cell_sbo_lib_2[["percent.mt"]] <- cell_library_2$percent.mt[cho_lib2_joint_bcs]

cho_cell_sbo_lib_2@meta.data$library <- rep("library 2", dim(cho_cell_sbo_lib_2)[2])

# add a new assay to hold the SBO data
cho_lib_2_sbo <- as.matrix(sbo_lib_2.data[, cho_lib2_joint_bcs]) # intersect

cho_cell_sbo_lib_2[["SBO"]] <- CreateAssayObject(counts = cho_lib_2_sbo)
```

### Classify
```{r}
# normalise
cho_cell_sbo_lib_2 <- NormalizeData(cho_cell_sbo_lib_2, assay = "SBO", 
                               normalization.method = "CLR")

eval_pq_lib2 = tibble()
for (i in seq(from=0.9, to=0.9999999, by =0.005)){

  cho_cell_sbo_lib_2 <- HTODemux(cho_cell_sbo_lib_2, assay = "SBO", 
                          positive.quantile = i)

  hto_demux_classification <- cho_cell_sbo_lib_2@meta.data %>%
    filter(SBO_classification.global != "Negative") %>%
    group_by(SBO_classification.global) %>%
    summarise(cells = n()) %>%
    mutate(threshold=i)
  
  eval_pq_lib2 <- bind_rows(eval_pq_lib2, hto_demux_classification) %>%
    mutate(library="library 2")
  }

eval_pq_lib2_plot <- eval_pq_lib2 %>%
  ggplot(aes(x=threshold, y=cells, color=SBO_classification.global)) +
           geom_point() +
           geom_line() + 
  geom_vline(xintercept = 0.975, linetype=11, size=0.2) +
  theme_bw() +
  labs(y="# Cells", x="HTODemux positive quantile treshold", color="SBO identification", title="SBO library 2")

cho_cell_sbo_lib_2 <- HTODemux(cho_cell_sbo_lib_2, assay = "SBO", 
                          positive.quantile = 0.975)

# show table
table(cho_cell_sbo_lib_2$SBO_maxID, cho_cell_sbo_lib_2$SBO_classification.global)

# data for merged classification plot
classification_library_2 <- data.frame(table(cho_cell_sbo_lib_2$SBO_maxID, cho_cell_sbo_lib_2$SBO_classification.global)) %>%
  mutate(library="RNA library 2")

# remove cells with no matched SBO 
cho_cell_sbo_lib_2.subset <- subset(cho_cell_sbo_lib_2, idents = "Negative", invert = T)
```

```{r}
figure_s3 <- (eval_pq_lib1_plot / eval_pq_lib2_plot) + plot_layout(guides = "collect") + plot_annotation(tag_levels = 'a') & theme(legend.position = 'bottom')

figure_s3
ggsave(plot = figure_s3, filename = paste(results_dir, "Figure S3.png", sep = ""), width = 5, height = 5, device = "png", dpi = 700)
```


## Combine cell library 1 & 2
```{r message=FALSE, warning=FALSE}
# merge non-negative cells
cho.combined.sbo <- merge(cho_cell_sbo_lib_1.subset,
  y = c(cho_cell_sbo_lib_2.subset),
  add.cell.ids = c("lib1", "lib2"),
  project = "TM_analysis", merge.data = TRUE
)

hto.dist.mtx.combined <- as.matrix(dist(t(GetAssayData(object = cho.combined.sbo, assay = "SBO"))))

cho.combined.sbo <- RunTSNE(cho.combined.sbo, distance.matrix = hto.dist.mtx.combined, 
                            perplexity = 100)

sbo_combined_tsne_plot <- DimPlot(cho.combined.sbo)

viridis_cp <- c("red", "#fde725", "#5ec962", "#21918c", "#31688e", "#440154")

figure_2d <- sbo_combined_tsne_plot$data %>%
  mutate(tag = case_when(
    ident == "8-hours-untreated-CCGGACTT" ~ "TM- [CCGGACTT]",
    ident == "1-hours-treated-GCTTGCCT" ~ "TM+ 1hr [GCTTGCCT]",
    ident == "2-hours-treated-GAACTCCC" ~ "TM+ 2hr [GAACTCCCT]",
    ident == "4-hours-treated-GTCAGGAT" ~ "TM+ 4hr [GTCAGGAT]",
    ident == "8-hours-treated-GTATATCC" ~ "TM+ 8hr [GTATATCC]",
    ident == "Doublet" ~ "Doublet",
  )) %>%
  ggplot(aes(x = tSNE_1, y = tSNE_2, color = tag)) +
  geom_point(size = 0.1) +
  theme_bw() +
  scale_color_manual(values = c("red", "#fde725", "#5ec962", "#21918c", "#31688e", "#440154")) +
  labs(x = "tSNE 1", y = "tSNE 2", color = "Sample [barcode]", title = "Combined SBO library 1 & 2") +
  guides(color = guide_legend(override.aes = list(size = 2)))

figure_2d
ggsave(plot = figure_2d, filename = paste(results_dir, "Figure_2C.png", sep = ""), width = 6, height = 4, device = "png", dpi = 700)

HTOHeatmap(cho.combined.sbo, assay = "SBO") + scale_fill_viridis_b()
```
### ridgeplot
```{r fig.height=3, fig.width=10}
Idents(cho.combined.sbo) <- "SBO_maxID"

tm_8hr_ridge <- RidgePlot(cho.combined.sbo, assay = "SBO", features = rownames(cho.combined.sbo[["SBO"]])[1], ncol = 1)$data %>%
  mutate(tag = "TM+ 8hr [GTATATCC]") %>%
  rename(expression = `8-hours-treated-GTATATCC`)

control_ridge <- RidgePlot(cho.combined.sbo, assay = "SBO", features = rownames(cho.combined.sbo[["SBO"]])[2], ncol = 1)$data %>%
  mutate(tag = "TM- [CCGGACTT]") %>%
  rename(expression = `8-hours-untreated-CCGGACTT`)

tm_4hr_ridge <- RidgePlot(cho.combined.sbo, assay = "SBO", features = rownames(cho.combined.sbo[["SBO"]])[3], ncol = 1)$data %>%
  mutate(tag = "TM+ 4hr [GTCAGGAT]") %>%
  rename(expression = `4-hours-treated-GTCAGGAT`)


tm_2hr_ridge <- RidgePlot(cho.combined.sbo, assay = "SBO", features = rownames(cho.combined.sbo[["SBO"]])[4], ncol = 1)$data %>%
  mutate(tag = "TM+ 2hr [GAACTCCCT]") %>%
  rename(expression = `2-hours-treated-GAACTCCC`)

tm_1hr_ridge <- RidgePlot(cho.combined.sbo, assay = "SBO", features = rownames(cho.combined.sbo[["SBO"]])[5], ncol = 1)$data %>%
  mutate(tag = "TM+ 1hr [GCTTGCCT]") %>%
  rename(expression = `1-hours-treated-GCTTGCCT`)

ridge_data <- bind_rows(control_ridge, tm_1hr_ridge, tm_2hr_ridge, tm_4hr_ridge, tm_8hr_ridge)

figure_s3 <- ridge_data %>%
  mutate(ident = case_when(
    ident == "8-hours-untreated-CCGGACTT" ~ "TM- [CCGGACTT]",
    ident == "1-hours-treated-GCTTGCCT" ~ "TM+ 1hr [GCTTGCCT]",
    ident == "2-hours-treated-GAACTCCC" ~ "TM+ 2hr [GAACTCCCT]",
    ident == "4-hours-treated-GTCAGGAT" ~ "TM+ 4hr [GTCAGGAT]",
    ident == "8-hours-treated-GTATATCC" ~ "TM+ 8hr [GTATATCC]",
    ident == "Doublet" ~ "Doublet",
  )) %>%
  arrange(desc(ident)) %>%
  ggplot(aes(x = expression, y = ident, fill = ident)) +
  geom_density_ridges(alpha = 0.7) +
  scale_y_discrete(expand = c(0, 0)) + # will generally have to set the `expand` option
  scale_x_continuous(expand = c(0, 0)) +
  scale_y_discrete(limits = rev) + # for both axes to remove unneeded padding
  coord_cartesian(clip = "off") + # to avoid clipping of the very top of the top ridgeline
  scale_fill_manual(values = c("#fde725", "#5ec962", "#21918c", "#31688e", "#440154")) +
  labs(x = "Expression level", y = "") +
  theme_ridges() +
  facet_wrap(~tag, nrow = 1) +
  theme(legend.position = "none")

figure_s3
ggsave(plot = figure_s3, filename = paste(results_dir, "Figure S3.png", sep = ""), width = 12, height = 4, device = "png", dpi = 700)
```
```{r}
sbo_classification_barplot <- bind_rows(cho_cell_sbo_lib_1@meta.data, cho_cell_sbo_lib_2@meta.data) %>%
  ggplot(aes(x = orig.ident, fill = SBO_classification.global)) +
  geom_bar(stat = "count", position = "stack") +
  labs(y = "# Cells", x = "", fill = "SBO \nClassification") +
  theme_bw() +
  theme(
    legend.position = "right",
    axis.text.x = element_text(angle = 35, vjust = 1, hjust = 1)
  ) +
  scale_fill_viridis(discrete = T, direction = -1, option = "plasma")

sbo_classification_barplot
ggsave(plot = sbo_classification_barplot, filename = paste(results_dir, "Figure_2B.png", sep = ""), 
       width = 2.5, height = 3, device = "png", dpi = 700)
```
### SBO classification by sample
```{r}
figure_s4 <- bind_rows(classification_library_1, classification_library_2) %>%
  mutate(ident = case_when(
    Var1 == "8-hours-untreated-CCGGACTT" ~ "TM- [CCGGACTT]",
    Var1 == "1-hours-treated-GCTTGCCT" ~ "TM+ 1hr [GCTTGCCT]",
    Var1 == "2-hours-treated-GAACTCCC" ~ "TM+ 2hr [GAACTCCCT]",
    Var1 == "4-hours-treated-GTCAGGAT" ~ "TM+ 4hr [GTCAGGAT]",
    Var1 == "8-hours-treated-GTATATCC" ~ "TM+ 8hr [GTATATCC]",
    Var1 == "Doublet" ~ "Doublet",
  )) %>%
  ggplot(aes(x=ident, y=Freq, fill=Var2)) +
  geom_bar(stat="identity",position = "stack") +
  labs(x="", y="# Cells", fill="SBO \nClassification") +
  facet_wrap(~library, nrow = 2) + 
  theme_bw() +
  theme(axis.text.x = element_text(angle = 35, vjust = 1, hjust = 1)) +
  scale_fill_viridis(discrete = T, direction = -1, option = "plasma")

figure_s4
ggsave(plot = figure_s4, filename = paste(results_dir, "Figure S4.png", sep = ""), width = 7, height = 5, device = "png", dpi = 700)
```

## Analysis of classified cells
```{r message=FALSE, warning=FALSE, include=FALSE}

# select only singlet cells from both libraries
Idents(cho_cell_sbo_lib_1) <- "SBO_classification.global"
cho_1.singlet <- subset(cho_cell_sbo_lib_1, idents = "Singlet")

Idents(cho_cell_sbo_lib_2) <- "SBO_classification.global"
cho_2.singlet <- subset(cho_cell_sbo_lib_2, idents = "Singlet")

#
cho.combined.sct <- merge(cho_1.singlet,
  y = c(cho_2.singlet),
  add.cell.ids = c("lib1", "lib2"),
  project = "TM_analysis", merge.data = TRUE
)

table(cho.combined.sct$SBO_maxID, cho.combined.sct$SBO_classification.global)

cho.combined.sct <- SCTransform(
  cho.combined.sct,
  assay = "RNA",
  new.assay.name = "SCT",
  vars.to.regress = c("nCount_RNA", "percent.mt")
)

cho.combined.sct <- CellCycleScoring(cho.combined.sct,
  assay = "SCT",
  s.features = cgr_s_phase,
  g2m.features = cgr_g2m_phase,
  set.ident = TRUE
)

cho.combined.sct <- SCTransform(
  cho.combined.sct,
  assay = "RNA",
  new.assay.name = "SCT",
  vars.to.regress = c("S.Score", "G2M.Score", "nCount_RNA", "percent.mt"),
  verbose = TRUE
)

sample_information <- cho.combined.sct@meta.data %>%
  mutate(treatment = case_when(
    SBO_maxID == "8-hours-untreated-CCGGACTT" ~ "control",
    SBO_maxID == "1-hours-treated-GCTTGCCT" ~ "TM+ 1hr",
    SBO_maxID == "2-hours-treated-GAACTCCC" ~ "TM+ 2hr",
    SBO_maxID == "4-hours-treated-GTCAGGAT" ~ "TM+ 4hr",
    SBO_maxID == "8-hours-treated-GTATATCC" ~ "TM+ 8hr",
  ))

cho.combined.sct <- AddMetaData(cho.combined.sct, metadata = sample_information$treatment, "treatment")
```

## Visualisation of tagged cells
```{r include=FALSE}
cho.combined.sct <- RunPCA(object = cho.combined.sct, verbose = FALSE, npcs = 50)
```

### Elbow plot
```{r fig.width=3}
figure_s5 <- ElbowPlot(cho.combined.sct,ndims = 100)

figure_s5 <- figure_s5$data %>%
  ggplot(aes(x=dims, y=stdev)) +
  geom_line() + 
  geom_point() +
  geom_vline(xintercept = 30, linetype=11, size=0.2, color="red") +
  theme_bw() +
  labs(x="Principal components", y="Standard Deviation")

figure_s5
ggsave(plot = figure_s5, filename = paste(results_dir, "Figure S3.png", sep = ""), width = 5, height = 5, device = "png", dpi = 700)
```


```{r include=FALSE}
cho.combined.sct <- RunUMAP(object = cho.combined.sct, dims = 1:40)
```

### Supplementary Figure 6
```{r fig.height=8, fig.width=4}
# overlay library

scale_fill_brewer(palette = "YlOrRd")
seurat_dim_plot_lib <- DimPlot(cho.combined.sct, reduction = "umap", group.by = c("library"), pt.size = 0.5)
# overlay cell cycle phase
seurat_dim_plot_cc <- DimPlot(cho.combined.sct, reduction = "umap", group.by = c("Phase"), pt.size = 0.5)

cbp1 <- c(
  "#E69F00", "#009E73",
  "#F0E442", "#0072B2", "#D55E00", "#CC79A7"
)

figure_s6a <- seurat_dim_plot_lib$data %>%
  ggplot(aes(x = UMAP_1, y = UMAP_2, color = library)) +
  geom_point(size = 0.1) +
  theme_bw() +
  scale_color_manual(values = cbp1) +
  labs(x = "UMAP 1", y = "UMAP 2", color = "Source") +
  theme(
    axis.ticks.x = element_blank(),
    axis.text.x = element_blank(),
    axis.ticks.y = element_blank(),
    axis.text.y = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()
  ) +
  guides(color = guide_legend(override.aes = list(size = 2)))

figure_s6b <- seurat_dim_plot_cc$data %>%
  ggplot(aes(x = UMAP_1, y = UMAP_2, color = Phase)) +
  geom_point(size = 0.1) +
  theme_bw() +
  scale_colour_brewer(palette = "Dark2",direction = 1) +
  # scale_color_manual(values = cbp1) +
  labs(x = "UMAP 1", y = "UMAP 2", color = "Phase") +
  theme(
    axis.ticks.x = element_blank(),
    axis.text.x = element_blank(),
    axis.ticks.y = element_blank(),
    axis.text.y = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()
  ) +
  guides(color = guide_legend(override.aes = list(size = 2)))


figure_s6 <- (figure_s6a / plot_spacer()+ figure_s6b) + plot_layout(heights =  c(4,0.5,4)) + plot_annotation(tag_levels = 'a')
figure_s6 
ggsave(plot = figure_s6, filename = paste(results_dir, "Figure S4.png", sep = ""), width = 4, height = 6.5, device = "png", dpi = 700)
```

### Plot classified cells
```{r}
sbo_tagged_cells_umap_plot <- DimPlot(cho.combined.sct, group.by = c("treatment"), combine = FALSE)

viridis_cp <- c("#fde725", "#5ec962", "#21918c", "#31688e", "#440154")


# figure_2ea <- sbo_tagged_cells_umap_plot[[1]]$data %>%
#   ggplot(aes(x = UMAP_1, y = UMAP_2)) +
#   geom_point(size = 0.1, color="grey") +
#   theme_bw() +
#   coord_fixed() +
#   scale_color_manual(values = viridis_cp) +
#   labs(x = "UMAP 1", y = "UMAP 2", color = "Treatment", subtitle = "") +
#   guides(color = guide_legend(override.aes = list(size = 2))) +
#   theme(
#     plot.title = element_text(size = 10, face = "bold.italic"),
#     legend.position = "right",
#     legend.title = element_text(size = 10),
#     legend.text = element_text(size = 9)
#   )

figure_2e <- sbo_tagged_cells_umap_plot[[1]]$data %>%
  ggplot(aes(x = UMAP_1, y = UMAP_2, color = treatment)) +
  geom_point(size = 0.1) +
  theme_bw() +
  coord_fixed() +
  scale_x_reverse() +
  #scale_color_brewer(palette = "Dark2",direction = 1) +
  scale_color_manual(values = viridis_cp) +
  labs(x = "UMAP 1", y = "UMAP 2", color = "Treatment", subtitle = "Classificaiton of cells by barcode") +
  guides(color = guide_legend(override.aes = list(size = 2))) +
  theme(
    plot.title = element_text(size = 10, face = "bold.italic"),
    legend.position = "right",
    legend.title = element_text(size = 10),
    legend.text = element_text(size = 9)
  )

figure_2e
ggsave(plot = figure_2e, filename = paste(results_dir, "Figure_2D.png", sep = ""), width = 5, height = 4, device = "png", dpi = 700)
```

### Cell cycle classification for each treatment
```{r}
figure_s7 <- cho.combined.sct@meta.data %>%
  group_by(treatment, Phase) %>%
  summarise(n = n()) %>%
  mutate(freq = n / sum(n)) %>%
  ggplot(aes(x = treatment, y = freq, fill = Phase)) +
  geom_bar(stat = "identity", position = "stack") +
  scale_fill_brewer(palette = "Dark2",direction = 1) +
  #scale_fill_manual(values = c("#CC6677", "#44AA99", "#88CCEE")) +
  theme_bw() +
  labs(x="",y = "Proportion of cells", fill = "Phase")

figure_s7
ggsave(plot = figure_s7 , filename = paste(results_dir, "Figure S7.png", sep = ""), width = 5, height = 3, device = "png", dpi = 700)
```
## Visualise known ER stress marker genes 
### UMAP
```{r}
hspa5_umap <- FeaturePlot(cho.combined.sct, features = c("Hspa5-1"))$data %>%
  mutate(gene = "Hspa5") %>%
  rename(expression = "Hspa5.1") %>%
  ggplot(aes(x = UMAP_1, y = UMAP_2, color = expression)) +
  geom_point(size = 0.01) +
  scale_color_viridis(option = "turbo", direction = 1) +
  theme_bw() +
  coord_fixed() +
  #scale_x_reverse() +
  labs(x = "UMAP 2", y = "UMAP 1", color = "Hspa5 \nExpression \nlevel") +
  theme(
    plot.title = element_text(size = 10, face = "bold.italic"),
    legend.position = "right",
    legend.title = element_text(size = 8),
    legend.text = element_text(size = 7)
  )

ddit3_umap <- FeaturePlot(cho.combined.sct, features = c("Ddit3-1"))$data %>%
  mutate(gene = "Ddit3") %>%
  rename(expression = "Ddit3.1") %>%
  ggplot(aes(x = UMAP_1, y = UMAP_2, color = expression)) +
  geom_point(size = 0.1) +
  scale_color_viridis(option = "turbo") +
  theme_bw() +
  coord_fixed() +
  #scale_x_reverse() +
  labs(x = "UMAP 2", y = "UMAP 1", color = "Ddit3 \nExpression \nlevel") +
  theme(
    plot.title = element_text(size = 10, face = "bold.italic"),
    legend.position = "right",
    legend.title = element_text(size = 8),
    legend.text = element_text(size = 7)
  )

atf4_umap <- FeaturePlot(cho.combined.sct, features = c("Atf4-1"))$data %>%
  mutate(gene = "Atf4") %>%
  rename(expression = "Atf4.1") %>%
  ggplot(aes(x = UMAP_1, y = UMAP_2, color = expression)) +
  geom_point(size = 0.1) +
  scale_color_viridis(option = "turbo") +
  theme_bw() +
  coord_fixed() +
  #scale_x_reverse() +
  labs(x = "UMAP 2", y = "UMAP 1", color = "Atf4 \nExpression \nlevel") +
  theme(
    plot.title = element_text(size = 10, face = "bold.italic"),
    legend.position = "right",
    legend.title = element_text(size = 8),
    legend.text = element_text(size = 7)
  )
```

#### Violin plots
```{r}
standard_normalisation <- NormalizeData(cho.combined.sct, normalization.method = "LogNormalize", scale.factor = 10000, assay = "RNA")

hspa5_violin <- VlnPlot(standard_normalisation, features = c("Hspa5-1"), group.by = "treatment", assay = "RNA")$data %>%
  mutate(gene = "Hspa5") %>%
  rename(expression = "Hspa5-1") %>%
  ggplot(aes(x = ident, y = `expression`, fill = ident)) +
  geom_violin() +
  geom_jitter(size = 0.1, alpha = 0.1, color = "black") +
  scale_fill_manual(values = c("#fde725", "#5ec962", "#21918c", "#31688e", "#440154")) +
  theme_bw() +
  # coord_fixed() +
  # ylim(0,3) +
  labs(x = "", y = "Expression Level", subtitle = "Hspa5") +
  theme(
    legend.position = "none",
    strip.text.x = element_text(face = "bold.italic"),
    strip.background = element_blank(),
    panel.spacing = unit(2, "lines"),
    axis.text.x = element_text(angle = 35, vjust = 1, hjust = 1)
  ) +
  scale_y_continuous(breaks = pretty_breaks(n = 4, min.n = 0))

atf4_violin <- VlnPlot(standard_normalisation, features = c("Atf4-1"), group.by = "treatment", assay = "RNA")$data %>%
  mutate(gene = "Atf4") %>%
  rename(expression = "Atf4-1") %>%
  ggplot(aes(x = ident, y = `expression`, fill = ident)) +
  geom_violin() +
  geom_jitter(size = 0.1, alpha = 0.1) +
  scale_fill_manual(values = c("#fde725", "#5ec962", "#21918c", "#31688e", "#440154")) +
  theme_bw() +
  # coord_fixed() +
  labs(x = "", y = "Expression Level", subtitle = "Atf4") +
  theme(
    legend.position = "none",
    strip.text.x = element_text(face = "bold.italic"),
    strip.background = element_blank(),
    panel.spacing = unit(2, "lines"),
    axis.text.x = element_text(angle = 35, vjust = 1, hjust = 1)
  ) +
  scale_y_continuous(breaks = pretty_breaks(n = 4, min.n = 0))

ddit3_violin <- VlnPlot(standard_normalisation, features = c("Ddit3-1"), group.by = "treatment", assay = "RNA")$data %>%
  mutate(gene = "Ddit3") %>%
  rename(expression = "Ddit3-1") %>%
  ggplot(aes(x = ident, y = `expression`, fill = ident)) +
  geom_violin() +
  geom_jitter(size = 0.1, alpha = 0.1) +
  # coord_fixed() +
  scale_fill_manual(values = c("#fde725", "#5ec962", "#21918c", "#31688e", "#440154")) +
  theme_bw() +
  labs(x = "", y = "Expression Level", subtitle = "Ddit3") +
  theme(
    legend.position = "none",
    strip.text.x = element_text(face = "bold.italic"),
    strip.background = element_blank(),
    axis.text.x = element_text(angle = 35, vjust = 1, hjust = 1)
  ) +
  scale_y_continuous(breaks = pretty_breaks(n = 4, min.n = 0))
```

### Manuscript Figure 
```{r}
marker_gene_umap <- (((hspa5_violin / hspa5_umap + plot_layout(heights = c(0.8, 1)))) |
  (atf4_violin / atf4_umap + plot_layout(heights = c(0.8, 1))) |
  (ddit3_violin / ddit3_umap + plot_layout(heights = c(0.7, 1))))

marker_gene_umap
ggsave(plot = marker_gene_umap, filename = paste(results_dir, "Figure 2E.png", sep = ""), width = 9, height = 4.5, device = "png", dpi = 700)
```


## Weighted gene coexpression network analysis
### Construct pseudobulk metacells
```{r}
genes.keep <- VariableFeatures(cho.combined.sct, assay = "SCT")

length(genes.keep)

seurat_list <- list()
for (group in unique(cho.combined.sct$treatment)) {
  print(group)
  cur_seurat <- subset(cho.combined.sct, treatment == group)
  cur_seurat <- cur_seurat[genes.keep, ]
  cur_metacell_seurat <- scWGCNA::construct_metacells(
    cur_seurat,
    name = group,
    k = 20,
    reduction = "umap",
    assay = "RNA", slot = "data"
  )
  seurat_list[[group]] <- cur_metacell_seurat
}

metacell_seurat <- merge(seurat_list[[1]], seurat_list[2:length(seurat_list)])

metacell_subset <- ScaleData(metacell_seurat, features = rownames(metacell_seurat))


metacell_subset <- RunPCA(metacell_subset, features = rownames(metacell_subset))
ElbowPlot(metacell_subset)


metacell_subset <- RunUMAP(metacell_subset, reduction = "pca", dims = 1:20)
metacell_umap <- DimPlot(metacell_subset, reduction = "umap", label = FALSE)
```

### Pseudobulk UMAP
```{r}
figure_3a <- metacell_umap[[1]]$data %>%
  ggplot(aes(x = UMAP_1, y = UMAP_2, color = ident)) +
  geom_point(size = 0.1) +
  theme_bw() +
  coord_fixed() +
  scale_color_manual(values = viridis_cp) +
  labs(x = "UMAP 1", y = "UMAP 2", color = "Treatment", subtitle = "") +
  guides(color = guide_legend(override.aes = list(size = 2))) +
  theme(
    plot.title = element_text(size = 10, face = "bold.italic"),
    legend.position = "right",
    legend.title = element_text(size = 10),
    legend.text = element_text(size = 9)
  )

figure_3a
ggsave(plot = figure_3a, filename = paste(results_dir, "Figure 3A.png", sep = ""), width = 5, height = 4, device = "png", dpi = 700)
```

```{r}
# enableWGCNAThreads()
nclusters <- length(unique(metacell_seurat$orig.ident))
genes.use <- rownames(metacell_seurat)
targets <- metacell_seurat@meta.data
group <- as.factor(metacell_seurat$orig.ident)
datExpr <- as.data.frame(GetAssayData(metacell_seurat, assay = "RNA", slot = "data")[genes.use, ])
datExpr <- as.data.frame(t(datExpr))
datExpr <- datExpr[, goodGenes(datExpr)]

# Choose a set of soft-thresholding powers
powers <- c(seq(1, 10, by = 1), seq(12, 50, by = 2))
# Call the network topology analysis function for each set in turn
powerTable <- list(
  data = pickSoftThreshold(
    datExpr,
    powerVector = powers,
    verbose = 100,
    networkType = "signed",
    corFnc = "bicor"
  )[[2]]
)
# Plot the results:
colors <- c("blue", "red", "black")
# Will plot these columns of the returned scale free analysis tables
plotCols <- c(2, 5, 6, 7)
colNames <- c(
  "Scale Free Topology Model Fit", "Mean connectivity", "mean connectivity",
  "Max connectivity"
)
# Get the minima and maxima of the plotted points
ylim <- matrix(NA, nrow = 2, ncol = 4)
for (col in 1:length(plotCols)) {
  ylim[1, col] <- min(ylim[1, col], powerTable$data[, plotCols[col]], na.rm = TRUE)
  ylim[2, col] <- max(ylim[2, col], powerTable$data[, plotCols[col]], na.rm = TRUE)
}

# Plot the quantities in the chosen columns vs. the soft thresholding power
par(mfcol = c(2, 2))
par(mar = c(4.2, 4.2, 2.2, 0.5))
cex1 <- 0.7
for (col in 1:length(plotCols)) {
  plot(powerTable$data[, 1], -sign(powerTable$data[, 3]) * powerTable$data[, 2],
    xlab = "Soft Threshold (power)", ylab = colNames[col], type = "n", ylim = ylim[, col],
    main = colNames[col]
  )
  addGrid()

  if (col == 1) {
    text(powerTable$data[, 1], -sign(powerTable$data[, 3]) * powerTable$data[, 2],
      labels = powers, cex = cex1, col = colors[1]
    )
  } else {
    text(powerTable$data[, 1], powerTable$data[, plotCols[col]],
      labels = powers, cex = cex1, col = colors[1]
    )
  }
  if (col == 1) {
    legend("bottomright", legend = "Metacells", col = colors, pch = 20)
  } else {
    legend("topright", legend = "Metacells", col = colors, pch = 20)
  }
}

powerTable$data %>%
  ggplot(aes(x=Power, y=SFT.R.sq)) +
  geom_point(size=0.1) +
  labs(y="Scale free topology fit", x="Power") +
  theme_bw()
  
powerTable$data %>%
  ggplot(aes(x=Power, y=mean.k.)) +
  geom_point(size=0.1) +
  labs(y="Mean connectivity", x="Power") +
  theme_bw()

powerTable$data %>%
  ggplot(aes(x=Power, y=mean.k.)) +
  geom_point(size=0.1) +
  labs(y="Mean connectivity", x="Power") +
  theme_bw() 
```
### Identify gene modules
```{r}
softPower <- 18

nSets <- 1
setLabels <- "treatment"
shortLabels <- setLabels

multiExpr <- list()
multiExpr[["TM"]] <- list(data = datExpr)

checkSets(multiExpr) # check data size

# construct network
net <- blockwiseConsensusModules(multiExpr,
  blocks = NULL,
  maxBlockSize = 30000, ## This should be set to a smaller size if the user has limited RAM
  randomSeed = 12345,
  corType = "pearson",
  power = softPower,
  consensusQuantile = 0.3,
  networkType = "signed",
  TOMType = "unsigned",
  TOMDenom = "min",
  scaleTOMs = TRUE, scaleQuantile = 0.8,
  sampleForScaling = TRUE, sampleForScalingFactor = 1000,
  useDiskCache = TRUE, chunkSize = NULL,
  deepSplit = 4,
  pamStage = FALSE,
  detectCutHeight = 0.995, 
  minModuleSize = 50,
  mergeCutHeight = 0.2,
  saveConsensusTOMs = TRUE,
  consensusTOMFilePattern = "ConsensusTOM-block.%b.rda"
)

consMEs <- net$multiMEs
moduleLabels <- net$colors
# Convert the numeric labels to color labels
moduleColors <- as.character(moduleLabels)
consTree <- net$dendrograms[[1]]
# module eigengenes
MEs <- moduleEigengenes(multiExpr[[1]]$data, colors = moduleColors, nPC = 1)$eigengenes
MEs <- orderMEs(MEs)
meInfo <- data.frame(rownames(datExpr), MEs)
colnames(meInfo)[1] <- "SampleID"

# intramodular connectivity
KMEs <- signedKME(datExpr, MEs, outputColumnName = "kME", corFnc = "bicor")

# compile into a module metadata table
geneInfo <- as.data.frame(cbind(colnames(datExpr), moduleColors, KMEs))

# geneInfo %>%
#   filter(Initially.Assigned.Module.Color == "black") %>%
#   arrange(-kMEblack)

# how many modules did we get?
nmodules <- length(unique(moduleColors))

# merged gene symbol column
colnames(geneInfo)[1] <- "GeneSymbol"
colnames(geneInfo)[2] <- "Initially.Assigned.Module.Color"

PCvalues <- MEs

png(file = paste(results_dir, "Figure_3A.png", sep = ""), units = "in", width = 6, height = 3, res = 700)

unique(moduleColors)

plotDendroAndColors(consTree, moduleColors, "Module colors",
  dendroLabels = FALSE, hang = 0.03, addGuide = TRUE, guideHang = 0.05,
  main = paste0("")
)

dev.off()

save(geneInfo, file="unannotated_gene_info.rData")

plotDendroAndColors(consTree, moduleColors, "Module colors",
  dendroLabels = FALSE, hang = 0.03, addGuide = TRUE, guideHang = 0.05,
  main = paste0("")
)

```

```{r}
 geneInfo %>%
   filter(Initially.Assigned.Module.Color == "magenta") %>%
   arrange(-kMEmagenta)
```


### Relationship between coexpressed modules and TM
```{r fig.height=10, fig.width=8}
plot_df <- cbind(select(targets, c(orig.ident)), PCvalues)
plot_df <- reshape2::melt(plot_df, id.vars = c("orig.ident"))
plot_df$orig.ident <- factor(plot_df$orig.ident, levels = c("control", "TM+ 1hr", "TM+ 2hr", "TM+ 4hr", "TM+ 8hr"))
colors <- sub("ME", "", as.character(levels(plot_df$variable)))

wgcna_boxplot_full <- plot_df %>%
  ggplot(aes(x = orig.ident, y = value, fill = orig.ident)) +
  scale_fill_manual(values = c("#fde725", "#5ec962", "#21918c", "#31688e", "#440154")) +
  geom_boxplot(notch = FALSE) +
  RotatedAxis() +
  ylab("Module Eigengene") +
  xlab("") +
  theme_bw() +
  theme(
    legend.position = "none",
    strip.text.x = element_text(face = "bold"),
    strip.background = element_blank(),
    axis.text.x = element_text(angle = 35, vjust = 1, hjust = 1)
  ) +
  facet_wrap(~variable, scales = "free", ncol = 3)


wgcna_boxplot_filtered <- plot_df %>%
  filter(variable %in% c("MEturquoise", "MEblue", "MEmagenta")) %>%
  mutate(new=factor(variable,labels=c("MEturquoise","MEblue","MEmagenta"),  levels=c("MEturquoise","MEblue","MEmagenta"))) %>%
  ggplot(aes(x = orig.ident, y = value, fill = orig.ident)) +
  scale_fill_manual(values = c("#fde725", "#5ec962", "#21918c", "#31688e", "#440154")) +
  geom_boxplot(notch = FALSE, width=0.4) +
  RotatedAxis() +
  ylab("Module Eigengene") +
  xlab("") +
  theme_bw() +
  theme(
    legend.position = "none",
    strip.text.x = element_text(face = "bold"),
    strip.background = element_blank(),
    axis.text.x = element_text(angle = 35, vjust = 1, hjust = 1),
    panel.spacing = unit(2, "lines")
  ) +
  facet_wrap(~new, scales = "free", ncol = 3)

wgcna_boxplot_filtered

ggsave(plot = wgcna_boxplot_filtered, filename = paste(results_dir, "Figure 3B.png", sep = ""), width = 9, height = 3.25, device = "png", dpi = 700)

geneInfo_annotated %>%
  filter(Initially.Assigned.Module.Color == "magenta")
```

### Annotate WGCNA results
```{r}
CriGri_PICRH_1_0_annotation <- read_delim(paste0("../data/GCF_003668045.3_CriGri-PICRH-1.0_feature_table.txt"),
  "\t",
  escape_double = FALSE, trim_ws = TRUE
)

geneInfo_annotated <- geneInfo %>%
  mutate(symbol = gsub("-.*", "", GeneSymbol)) %>%
  left_join(CriGri_PICRH_1_0_annotation, by = "symbol") %>%
  filter(`# feature` == "mRNA") %>%
  mutate(name = gsub(",.*", "", name)) %>%
  distinct(GeneSymbol, .keep_all = T) %>%
  dplyr::select("symbol", "name", "GeneID", c(colnames(geneInfo)))


pcg_without_loc_ids <- geneInfo_annotated %>%
  filter(!str_detect(symbol, "LOC"))

pcg_with_loc_ids <- geneInfo_annotated %>%
  filter(str_detect(symbol, "LOC"))


mouse_feature_table <- read_delim("../data/GCF_000001635.27_GRCm39_feature_table.txt",
  delim = "\t", escape_double = FALSE,
  trim_ws = TRUE
)

loc_id_gene_symbol <- mouse_feature_table %>%
  mutate(ncbi_symbol = symbol) %>%
  fuzzy_inner_join(pcg_with_loc_ids, by = "name", match_fun = str_detect) %>%
  distinct(symbol.y, .keep_all = T)

pcg_with_loc_ids <- pcg_with_loc_ids %>%
  mutate("symbol.y" = symbol) %>%
  left_join(loc_id_gene_symbol, by = "symbol.y") %>%
  mutate(ncbi_symbol = coalesce(ncbi_symbol, symbol.y)) %>%
  mutate(
    "symbol" = ncbi_symbol,
    "GeneSymbol" = GeneSymbol.x,
    "Initially.Assigned.Module.Color" = Initially.Assigned.Module.Color.x,
    "kMEyellow" = kMEyellow.x,
    "kMEpurple" = kMEpurple.x,
    "kMEblack" = kMEblack.x,
    "kMEmagenta" = kMEyellow.x,
    "kMEpink" = kMEpink.x,
    "kMEturquoise" = kMEturquoise.x,
    "kMEred" = kMEred.x,
    "kMEgreen" = kMEgreen.x,
    "kMEblue" = kMEblue.x,
    "kMEbrown" = kMEbrown.x,
    "kMEgrey" = kMEgrey.x
  ) %>%
  select(symbol, GeneSymbol, c(colnames(geneInfo_annotated)))

geneInfo_annotated <- bind_rows(pcg_without_loc_ids, pcg_with_loc_ids)


# save info
write.csv(geneInfo_annotated, file = paste0(results_dir, "WGCNA_geneInfo_signed.csv"))
```




```{r}
geneInfo %>%
  filter(Initially.Assigned.Module.Color == "turquoise") %>%
  arrange(-kMEbrown)

goi <- "Insig1-1"

test_umap <- FeaturePlot(cho.combined.sct, features = c(goi))$data %>%
  mutate(gene = "goi") %>%
  rename(expression = "Insig1.1") %>%
  ggplot(aes(x = UMAP_1, y = UMAP_2, color = expression)) +
  geom_point(size = 0.1) +
  scale_color_viridis(option = "turbo", direction = 1) +
  theme_bw() +
  coord_fixed() +
  labs(x = "UMAP 2", y = "UMAP 1", color = "Hspa5 \nExpression \nlevel") +
  theme(
    plot.title = element_text(size = 10, face = "bold.italic"),
    legend.position = "right",
    legend.title = element_text(size = 8),
    legend.text = element_text(size = 7)
  )

test_umap + figure_2e
```


### Gene set enrichment analysis
```{r}
enrichdir <- paste0(results_dir, "/enrichment_analysis/")
suppressMessages(if (file.exists(enrichdir)) {
  unlink(enrichdir, recursive = T)
})
if (!dir.exists(enrichdir)) {
  dir.create(enrichdir)
}


selected_colors <- c("turquoise", "blue", "magenta")
for (color in selected_colors) {
  current_genes <- geneInfo_annotated %>%
    filter(Initially.Assigned.Module.Color == color) %>%
    select(symbol)

  write(current_genes$symbol, file = paste0(enrichdir, color, "_genes.txt"))
}

enrich_result <- WebGestaltRBatch(
  enrichMethod = "ORA",
  organism = "mmusculus",
  enrichDatabase = c("geneontology_Biological_Process_noRedundant"),
  enrichDatabaseType = "genesymbol",
  interestGeneFolder = enrichdir,
  interestGeneType = "genesymbol",
  referenceSet = "genome",
  minNum = 10,
  maxNum = 500,
  sigMethod = "fdr",
  fdrMethod = "BH",
  fdrThr = 0.05,
  topThr = 10,
  reportNum = 20,
  perNum = 1000,
  projectName = "SBO ER stress",
  isOutput = TRUE,
  outputDirectory = enrichdir,
  dagColor = "continuous",
  setCoverNum = 10,
  networkConstructionMethod = NULL,
  neighborNum = 10,
  highlightType = "Seeds",
  highlightSeedNum = 10,
  nThreads = 32
)
```


```{r fig.height=9, fig.width=10}
blue_module_enrichment <- enrich_result[[1]]$enrichResult[, c(1, 2, 4, 5, 7, 8, 9, 11)] %>%
  mutate(type = "MEblue")

magenta_module_enrichment <- enrich_result[[2]]$enrichResult[, c(1, 2, 4, 5, 7, 8, 9, 11)] %>%
  mutate(type = "MEmagenta")

turquoise_module_enrichment <- enrich_result[[3]]$enrichResult[, c(1, 2, 4, 5, 7, 8, 9, 11)] %>%
  mutate(type = "MEturquoise")


enrichment_plot_data <- bind_rows(turquoise_module_enrichment[1:6, ], blue_module_enrichment[1:6, ], magenta_module_enrichment[1:6, ])

enrichment_plots <- enrichment_plot_data %>%
  mutate(FDR = case_when(FDR == 0 ~ 2.2e-16, TRUE ~ FDR)) %>%
  mutate(new_description = str_wrap(description, width = 20)) %>%
  mutate(new=factor(type,labels=c("MEturquoise","MEblue","MEmagenta"),  levels=c("MEturquoise","MEblue","MEmagenta"))) %>%
   mutate(new_order = fct_reorder(new_description, FDR)) %>%
  arrange(new_description, FDR) %>%
  ggplot(aes(y = -log10(FDR), x =new_description, fill = new, label=overlap)) +
  geom_bar(stat="identity", width = 0.4) +
  #scale_y_continuous(expand = c(0, 0), limits = c(0, -log10(FDR)+1)) +
  geom_text(aes(label = overlap), vjust = 0.5,hjust = 1.1, colour = "white") +
  #scale_fill_viridis() +
  theme_bw() +
  labs(x = "GO Biological Process",y = "-log10(FDR)") +
  facet_wrap(~new, scales = "free", ncol = 3) +
  scale_fill_manual(values=c("turquoise","blue", "magenta")) +
  theme(strip.text.x = element_text(face = "bold"),strip.background = element_blank(), legend.position = "none",
        axis.text.x = element_text( size = 6),
        panel.spacing = unit(2, "lines")) +
    scale_x_discrete(limits=rev) +
  coord_flip()

test <- wgcna_boxplot_filtered / enrichment_plots   + plot_layout(heights = c(1, 1.5))
ggsave(plot = test, filename = paste(results_dir, "Figure 3C.png", sep = ""), width = 10, height =6.5, device = "png", dpi = 700)
```

```{r fig.width=4}
library(reshape2)
library(sna)
library(Hmisc)
library(ggnetwork)
library(gplots)


numgenesingraph = 25;
numconnections2keep = 500;

TOM <- as.matrix(consTomDS)
geneInfo=geneInfo[geneInfo$GeneSymbol!="NA",]

mod ="black"

colind = which(colnames(geneInfo)==paste('kME',mod,sep=''));
rowind = which(geneInfo$Initially.Assigned.Module.Color==mod);

cat(' ',length(rowind),'probes in module\n'); 

submatrix = geneInfo[rowind,];
orderind = order(submatrix[,colind],decreasing=TRUE);

submatrix = submatrix[orderind[1:numgenesingraph],];
matchind = match(submatrix$GeneSymbol,colnames(datExpr));
reducedTOM = TOM.matrix[matchind,matchind];

dim(reducedTOM)

dimnames(reducedTOM) <- list(geneInfo$GeneSymbol[matchind],geneInfo$GeneSymbol[matchind])

netdf <- melt(lower.tri.remove(reducedTOM)) # matrix is symmetrical, we designate half of it at NA with lower.tri.remove
names(netdf)[1:2] <- c("gene1","gene2")
netdf <- netdf[complete.cases(netdf),]
netdf_sub <- subset(netdf, value > 0 & gene1!=gene2)
netdf_sub <- subset(netdf_sub, value > quantile(netdf_sub$value,0.95))

# netdf_sub <- subset(netdf_sub, gene1 %nin% names(net$colors)[which(net$colors=="grey" | net$colors=="yellow" | net$colors=="green" | net$colors=="blue" | net$colors=="red" | net$colors=="brown" | net$colors=="black")] & gene2 %nin% names(net$colors)[which(net$colors=="grey" | net$colors=="yellow" | net$colors=="green" | net$colors=="blue" | net$colors=="red" | net$colors=="brown" | net$colors=="black")])



edges <- netdf_sub[,c("gene1","gene2")]
edges$gene1 <- as.character(edges$gene1)
edges$gene2 <- as.character(edges$gene2)

edges_forplot <- network(edges,directed=F,loops=FALSE)
x <- data.frame(vertices=network.vertex.names(edges_forplot))
set.edge.attribute(edges_forplot,"similarity", netdf_sub$value)


n <- ggnetwork(edges_forplot,  niter = 100, cell.jitter = 0.75)
n$module <- net$colors[match(n$vertex.names,names(net$colors))]

ucols <- unique(n$module)
uhex <-  col2hex(ucols)
colpal <- uhex
names(colpal) <- ucols

vertex_symbols <- n %>%
  mutate(GeneSymbol=vertex.names) %>%
left_join(geneInfo_annotated, n, by="GeneSymbol") %>%
  select(symbol)


network_turquoise <- ggplot(n, aes(x = x, y = y, xend = xend, yend = yend)) +
  geom_edges(size=0.3,curvature=0.1,col="grey85",show.legend = F) +
  geom_nodes(aes(colour=module, size=0.2)) +
  geom_nodelabel(aes(label = vertex_symbols$symbol, colour=module),label.size = 0.50
                ) +
  #scale_size_continuous(range=c(0.1,3))+
  scale_colour_manual(values=colpal)+
  theme_blank()+
  theme(legend.position = "none")
network_turquoise
ggsave(plot = network_turquoise, filename = paste(results_dir, "Figure 3F.png", sep = ""), width = 7, height = 4.5, device = "png", dpi = 700)

merge(geneInfo_annotated, n$vertex.names)
a <- geneInfo_annotated$symbol[geneInfo_annotated$GeneSymbol %in% n$vertex.names] 

```

```{r}
geneInfo_annotated %>%
  filter(Initially.Assigned.Module.Color == "blue") %>%
  arrange(-kMEblue)
```



```{r}

uso1_umap <- FeaturePlot(cho.combined.sct, features = c("Uso1-1"))$data %>%
  mutate(gene = "Uso1") %>%
  rename(expression = "Uso1.1") %>%
  ggplot(aes(x = UMAP_1, y = UMAP_2, color = expression)) +
  geom_point(size = 0.01) +
  scale_color_viridis(option = "turbo", direction = 1) +
  theme_bw() +
  coord_fixed() +
  #scale_x_reverse() +
  labs(x = "UMAP 2", y = "UMAP 1", color = "Uso1 \nExpression \nlevel") +
  theme(
    plot.title = element_text(size = 10, face = "bold.italic"),
    legend.position = "right",
    legend.title = element_text(size = 8),
    legend.text = element_text(size = 7)
  )

asap1_umap <- FeaturePlot(cho.combined.sct, features = c("Asap1-1"))$data %>%
  mutate(gene = "Asap1") %>%
  rename(expression = "Asap1.1") %>%
  ggplot(aes(x = UMAP_1, y = UMAP_2, color = expression)) +
  geom_point(size = 0.01) +
  scale_color_viridis(option = "turbo", direction = 1) +
  theme_bw() +
  coord_fixed() +
  #scale_x_reverse() +
  labs(x = "UMAP 2", y = "UMAP 1", color = "Asap1 \nExpression \nlevel") +
  theme(
    plot.title = element_text(size = 10, face = "bold.italic"),
    legend.position = "right",
    legend.title = element_text(size = 8),
    legend.text = element_text(size = 7)
  )



hmgcs1_umap <- FeaturePlot(cho.combined.sct, features = c("Hmgcs1-1"))$data %>%
  mutate(gene = "Hmgcs1") %>%
  rename(expression = "Hmgcs1.1") %>%
  ggplot(aes(x = UMAP_1, y = UMAP_2, color = expression)) +
  geom_point(size = 0.01) +
  scale_color_viridis(option = "turbo", direction = 1) +
  theme_bw() +
  coord_fixed() +
  #scale_x_reverse() +
  labs(x = "UMAP 2", y = "UMAP 1", color = "Hmgcs1 \nExpression \nlevel") +
  theme(
    plot.title = element_text(size = 10, face = "bold.italic"),
    legend.position = "right",
    legend.title = element_text(size = 8),
    legend.text = element_text(size = 7)
  )

uso1_umap + asap1_umap + hmgcs1_umap

ggsave(filename = paste(results_dir, "Figure 3test.png", sep = ""), width = 10, height = 3.5, device = "png", dpi = 700)
```

